src/diffusers/models/adapter.py

```diff
@@ -0,0 +1,246 @@
+        else:
+            # if block_mid_channels[i] < block_out_channels[i](bottleneck downsample block), using light weight adapter schema instead
+            self.conv_in = nn.Conv2d(
+                channels_in * input_scale_factor**2,
+                block_mid_channels[0],
+                proj_kerenl_size,
+                1,
+                proj_kerenl_size // 2,
+            )
+
+    def forward(self, x: torch.Tensor) -> List[torch.Tensor]:
+        r"""
+        Args:
+            x (`torch.Tensor`):
+                (batch, channel, height, width) input images for adapter model, `channel` should equal to
+                `channels_in`.
+        """
+        # unshuffle
+        x = self.unshuffle(x)
+        # extract features
+        features = []
+        x = self.conv_in(x)
+        for i in range(self.num_downsample_blocks):
+            for j in range(self.num_res_blocks):
+                idx = i * self.num_res_blocks + j
+                x = self.body[idx](x)
+            features.append(x)
+
+        return features
+
+
+class MultiAdapter(ModelMixin):
+    r"""
+    MultiAdapter is a wrapper model that contains multiple adapter models and merges their outputs according to
+    user-assigned weighting.
+
+    This model inherits from [`ModelMixin`]. Check the superclass documentation for the generic methods the library
+    implements for all the model (such as downloading or saving, etc.)
+
+    Parameters:
+        adapters (`List[Adapter]`, *optional*, defaults to None):
+            A list of `Adapter` model instances.
+    """
+
+    def __init__(self, adapters: List[Adapter]):
+        super(MultiAdapter, self).__init__()
+
+        self.num_adapter = len(adapters)
+        self.adapters = nn.ModuleList(adapters)
+
+    def forward(self, xs: torch.Tensor, adapter_weights: Optional[List[float]] = None) -> List[torch.Tensor]:
+        r"""
+        Args:
+            xs (`torch.Tensor`):
+                (batch, channel, height, width) input images for multiple adapter models concated along dimension 1,
+                `channel` should equal to `num_adapter` * "number of channel of image".
+            adapter_weights (`List[float]`, *optional*, defaults to None):
+                List of floats representing the weight which will be multiply to each adapter's output before adding
+                them together.
+        """
+        if adapter_weights is None:
+            adapter_weights = torch.tensor([1 / self.num_adapter] * self.num_adapter)
+        else:
+            adapter_weights = torch.tensor(adapter_weights)
+
+        if xs.shape[1] % self.num_adapter != 0:
+            raise ValueError(
+                f"Expecting multi-adapter's input have number of channel that cab be evenly divisible "
+                f"by num_adapter: {xs.shape[1]} % {self.num_adapter} != 0"
+            )
+        x_list = torch.chunk(xs, self.num_adapter, dim=1)
```

Comment:
Interesting! 

I would have thought that we pass `xs` to the individual adapters and then weigh them. But it seems like we chunk `xs` w.r.t the channel dim and pass the chunks to the adapters in a sequential manner. 

Just ensuring my understanding is correct.

Reply:
The pipeline will stack the control images of each adapter into a single tensor with the shape of `(batch_size, num_adapter * 3, h, w)` after preprocessing, and this tensor will be the `xs` when it reaches `MultiAdapter` forward-pass method. So the chunk operation is essentially breaking down `xs` back to its source control images, sending them to corresponding adapters, and then weighting sum all the adapter outputs. 
We are doing this so that `MultiAdapter` behaves the same as a normal Adapter to the outside world (pipeline), by accepting the same type of parameters.